tasks/tasks.go

```diff
@@ -0,0 +1,577 @@
+	if len(tasks) > 0 {
+		slog.Debug(fmt.Sprintf("Saving %d tasks to %s", len(tasks), dataFile))
+		file, err := os.OpenFile(dataFile, os.O_RDWR|os.O_CREATE, 0644)
+		if err != nil {
+			return fmt.Errorf("Opening task manager file %s: %s", dataFile, err.Error())
+		}
+		enc := gob.NewEncoder(file)
+		err = enc.Encode(tasks)
+		if err != nil {
+			file.Close()
+			os.Remove(dataFile)
+			return fmt.Errorf("Saving tasks: %s", err.Error())
+		}
+		err = file.Close()
+		if err != nil {
+			os.Remove(dataFile)
+			return fmt.Errorf("Writing task manager file %s: %s", dataFile, err.Error())
+		}
+		slog.Debug(fmt.Sprintf("Saved %d tasks to %s", len(tasks), dataFile))
+	}
+	return nil
+}
+
+// this type holds various channels used by the TaskManager to communicate
+// with its worker goroutine
+type channelsType struct {
+	TaskId chan uuid.UUID      // task ID channel
+	Task   chan taskType       // task input channel
+	Status chan TransferStatus // status output channel
+	Error  chan error          // error output channel
+	Poll   chan struct{}       // polling channel
+	Stop   chan struct{}       // stop channel
+}
+
+// this function runs in its own goroutine, using the given local endpoint
+// for local file transfers, and the given channels to communicate with
+// the TaskManager
+func processTasks() {
+	// create or recreate a persistent table of transfer-related tasks
+	dataStore := filepath.Join(config.Service.DataDirectory, "dts.gob")
+	tasks := createOrLoadTasks(dataStore)
+
+	// parse the task channels into directional types as needed
+	var taskIdChan chan uuid.UUID = taskChannels.TaskId
+	var taskChan <-chan taskType = taskChannels.Task
+	var statusChan chan<- TransferStatus = taskChannels.Status
+	var errorChan chan<- error = taskChannels.Error
+	var pollChan <-chan struct{} = taskChannels.Poll
+	var stopChan <-chan struct{} = taskChannels.Stop
+
+	// the task deletion period is specified in seconds
+	deleteAfter := time.Duration(config.Service.DeleteAfter) * time.Second
+
+	// start scurrying around
+	for {
+		select {
+		case newTask := <-taskChan: // Add() called
+			newTask.Id = uuid.New()
+			tasks[newTask.Id] = newTask
+			taskIdChan <- newTask.Id
+			slog.Info(fmt.Sprintf("Added new transfer task %s.", newTask.Id.String()))
+		case taskId := <-taskIdChan: // Status() called
+			if task, found := tasks[taskId]; found {
+				statusChan <- task.Status
+			} else {
+				err := fmt.Errorf("Task %s not found!", taskId.String())
+				errorChan <- err
+			}
+		case <-pollChan: // time to move things along
+			for taskId, task := range tasks {
+				oldStatus := task.Status
+				err := task.Update()
+				if err != nil {
+					// we log task update errors but do not propagate them
+					slog.Error(err.Error())
+				}
+				if task.Status.Code != oldStatus.Code {
+					switch task.Status.Code {
+					case TransferStatusStaging:
+						slog.Info(fmt.Sprintf("Staging files for task %s.", task.Id.String()))
```

Comment:
this is just a note, but I find when I'm writing logging stuff (or error/output messages) to structure them such that the ID is always in the same place - makes the logs look more uniform and it's easier to find info about a specific task of interest. E.g.

```
Task 12345: beginning transfer
Task 12345: suspended transfer
Task 12345: staging files
Task 23456: purging
Task 93847: task not found
```

vs

```
beginning transfer for Task 12345
suspended transfer for Task 12345
staging files for Task 12345
purging Task 23456 
Task 93847 not found
```

Reply:
I can do this in a cleanup PR (#43).